packages/agent/src/tool/read-file/ReadFileTool.ts

```diff
@@ -50,12 +50,22 @@ export const executeReadFile =
     workspacePath: string;
   }): ExecuteToolFunction<ReadFileInput, ReadFileOutput> =>
   async ({ input: { filePath } }) => {
-    // TODO try-catch
-    const fullPath = path.join(workspacePath, filePath);
-    const content = await fs.readFile(fullPath, "utf-8");
+    try {
+      const fullPath = path.join(workspacePath, filePath);
+      const content = await fs.readFile(fullPath, "utf-8");
 
-    return {
-      summary: `Read file ${filePath}`,
-      output: { content },
-    };
+      return {
+        summary: `Read file ${filePath}`,
+        output: { content },
+      };
+    } catch (error: any) {
+      console.log(error);
+      if (error.code === "ENOENT") {
+        const files = await fs.readdir(workspacePath);
+        const availableFiles = files.join(", ");
+        throw new Error(`File not found. Available files: ${availableFiles}`);
```

Comment:
ReadFileTool will be used for deep paths (e.g. `src/something/something`). Listing all files could lead to prompt size issues. I like the idea of listing the available files, but I'm not sure how practical it'll be. For the dev agent prototype I worked on last month I had a dedicated tool to list the files (and exclude e.g. gitignored files).

Reply:
I noticed that this action sometimes fails and GPT opts to simply try reading it again and then gives up. So I believe listing files in the directory of the listed file would likely prompt it into a more usable direction while also saving a step or two because it ideally even directly has more useful information.

This is not doing (and not intended to do) a deep search.

The assumption is if you don't find the file you'll almost always want to list the paths so why waste the time and possibly get the thing stuck when you can give the relevant info right ahead.

Reply:
Yup, I fully understand. I really like the core idea of returning a list of relevant files and see how it is helpful. It's just that it'll return irrelevant files when you try to find a file in a deeper path (which is the way more common scenario in any software development repo), and that can lead the LLM down the wrong path.